File: tests/examples/namespace.cpp
#include <string>

namespace ExNS {
class ExNSClz {};
};  // namespace ExNS

class ExNsClass {
    ExNS::ExNSClz m;
};

namespace abc {
class Ns {
public:
    int a;
    class Inner {
        int b;
    };
    struct InSt {
        int c;
    };
    Inner i;
    InSt s;
};

enum NsEnum { A, B, C };
};  // namespace abc

namespace a {
namespace b {
class c {
    class d {};
};
};  // namespace b
};  // namespace a

namespace def {
typedef struct {
    int i;
} NsTypeDefSt;

};  // namespace def

class Outer {
public:
    class In {
        int b;
    };
    In i;
};

class StOuter {
public:
    struct StIn {
        int b;
    } si;
};

class Outer2 {
private:
    class In2;
    static In2 i;
};
class Outer2::In2 {
public:
    explicit In2(int i) : b(i) {}

private:
    int b;
};
Outer2::In2 Outer2::i(0);

class StInClass {
    typedef struct {
        int a;
    } InClassSt;

    InClassSt ast;
};

class NsClass {
    abc::Ns m;
};

class NsEnumClass {
    abc::NsEnum e;
};

class NsInnerClass {
    abc::Ns::Inner i;
};

class NsInnerStruct {
    abc::Ns::InSt s;
};

class NsInnerTypedef {
    def::NsTypeDefSt t;
};

class MyString : public std::string {};

File: tests/examples/namespace.py
from examples.utils import *

def case_namespace():
    data = read_json('namespace.cpp.json')
    expect_eq(data['NsClass']['fields'][0]['types'][0]['isReduced'] , True)
    expect_eq(data['NsClass']['isInner'] , False)
    expect_eq(data['abc::Ns']['isInner'] , False)
    expect_eq(data['abc::Ns::Inner']['isInner'] , True)
    expect_eq(data['abc::NsEnum']['isInner'] , False)
    expect_eq(data['NsInnerClass']['fields'][0]['types'][0]['name'] , 'abc::Ns::Inner')
    expect_eq(data['abc::Ns::InSt']['isInner'] , True)
    expect_eq(data['NsInnerStruct']['fields'][0]['types'][0]['name'] , 'abc::Ns::InSt')
    expect_true(any((field["name"] == "i" and field["types"][0]["name"] == "abc::Ns::Inner" for field in data["abc::Ns"]["fields"])))
    expect_eq(data['Outer']['fields'][0]['types'][0]['name'] , 'Outer::In')
    expect_eq(data['Outer2']['fields'][0]['types'][0]['name'] , 'Outer2::In2')
